### PR TITLE
Lazy initialisation & caching of scroll request, fuzzy scroll name search

### PR DESCRIPTION
**Pull request by ShareDVI**
Solves #9 
Adds the ability to do fuzzy scroll name search (i.e. searchs with typos)

The semantics of scrollsguide.py is now changed a bit.

Haven't tested it, sorry - I think I need to go to discord to get a key...

### CODE REVIEW

File: src/communication/scrollsguide.py [hunk 1]
```diff
@@ -4,10 +4,29 @@
...
+
+
+class MultipleScrollsFound(Exception):
+    def __init__(self, scrolls, search_term):
+        self.scrolls = [scroll.name for scroll in scrolls]
+        self.search_term = search_term
+
+    def __str__(self):
+        if len(self.scrolls) > 5:
+            return f"Too many scrolls start with {self.search_term}"
```
Comment (iScrE4m) [inline src/communication/scrollsguide.py]: These two return strings should maybe say that they're doing a proper search, not just "starting with"

File: src/cogs/game.py [hunk 2]
```diff
@@ -11,11 +11,11 @@ class Game:
     def __init__(self, bot: commands.bot.Bot):
         self.bot = bot
 
-    @commands.command(decsription="Get info about a scroll", brief="Get info about a scroll")
+    @commands.command(description="Get info about a scroll", brief="Get info about a scroll")
```
Comment (iScrE4m) [inline src/cogs/game.py]: Embarassing, thanks!

File: src/communication/scrollsguide.py [hunk 3]
```diff
@@ -87,37 +106,42 @@ def types(self) -> str:
...
+                data = json.loads(text)
+                cls._scrolls_db = {}
+                cls._scrolls_names = {}
+                for scroll_data in data.get('data', []):
+                    scroll = Scroll(scroll_data)
+                    cls._scrolls_db[scroll._id] = scroll
+                    cls._scrolls_names[scroll.name.lower()] = scroll._id
+
+    @classmethod
+    async def get(cls: Type['Scroll'], query: str, threshold: float = 0.7) -> 'Scroll':
```
Comment (iScrE4m) [inline src/communication/scrollsguide.py]: Let's call it `get_by_name`, `from_name`, `by_name` or something like that :)

File: src/communication/scrollsguide.py [hunk 4]
```diff
@@ -4,10 +4,29 @@
...
+    def __str__(self):
+        if len(self.scrolls) > 5:
+            return f"Too many scrolls start with {self.search_term}"
+        return f'Multiple scrolls starting with {self.search_term}: {", ".join(self.scrolls)}'
 
 class Scroll:
     """Wrapper class for Scroll information coming from the API"""
+
+    _scrolls_db = None
+    _scrolls_names = None
```
Comment (iScrE4m) [inline src/communication/scrollsguide.py]: Do we need both? If you just keep names and put entire Scroll in there, it shouldn't cause much troubles right?

File: src/communication/scrollsguide.py [hunk 5]
```diff
@@ -87,37 +106,42 @@ def types(self) -> str:
...
-                    return Scroll(scroll)
-                elif scroll['name'].lower().startswith(name.lower()):
-                    backup_scrolls.append(scroll)
-            if len(backup_scrolls) == 1:
-                return Scroll(backup_scrolls[0])
-            elif backup_scrolls:
-                raise MultipleScrollsFound(backup_scrolls, name)
-            raise ScrollNotFound
+    @classmethod
+    async def load_scrolls(cls: Type['Scroll']) -> None:
```
Comment (iScrE4m) [inline src/communication/scrollsguide.py]: There's an argument for load scrolls to be `__private`. I don't know how strong of an argument myself, but there is one :D

File: src/communication/scrollsguide.py [hunk 6]
```diff
@@ -87,37 +106,42 @@ def types(self) -> str:
...
+        else:
+            # do a fuzzy search
+            name_scores = {}
+            for scroll_name, id in cls._scrolls_names.items():
+                # 0.995 and 1.000 are really the same for grouping
+                score = round(fuzz.partial_ratio(query, scroll_name), 2)
+                if score >= threshold:
+                    name_scores[id] = score
+                # Only select top items from the list
+                max_score = max(name_scores.values())
```
Comment (iScrE4m) [inline src/communication/scrollsguide.py]: Shouldn't this be unindented a fter the for loop?

File: src/communication/scrollsguide.py (hunk 1 above)
Comment (ShareDVI) [reply src/communication/scrollsguide.py]: Agreed

File: src/communication/scrollsguide.py (hunk 3 above)
Comment (ShareDVI) [reply src/communication/scrollsguide.py]: Ok

File: src/communication/scrollsguide.py (hunk 4 above)
Comment (ShareDVI) [reply src/communication/scrollsguide.py]: We could. But what if we need to store scroll id somewhere and then by id. But sure, fair enough

File: src/communication/scrollsguide.py (hunk 5 above)
Comment (ShareDVI) [reply src/communication/scrollsguide.py]: I was thinking of a possible later command to reload scrolls manually, but as it is, it's more logical to be private, alright

File: src/communication/scrollsguide.py (hunk 6 above)
Comment (ShareDVI) [reply src/communication/scrollsguide.py]: Duh! you're right, the last line should.

**Comment by ShareDVI** [comment]
ok, it's fixed

**Pull request merged**